modules/network/src/main/java/org/apache/ignite/network/scalecube/ScaleCubeTopologyService.java

```diff
@@ -30,49 +32,62 @@
  * Implementation of {@link TopologyService} based on ScaleCube.
  */
 final class ScaleCubeTopologyService extends AbstractTopologyService {
-    /** Inner representation a ScaleCube cluster. */
-    private Cluster cluster;
+    /** Local member node. */
+    private ClusterNode localMember;
+
+    /** Topology members. */
+    private Map<String, ClusterNode> members = new HashMap<>();
 
     /**
-     * Sets the ScaleCube's {@link Cluster}. Needed for cyclic dependency injection.
+     * Sets the ScaleCube's local {@link Member}.
      */
-    void setCluster(Cluster cluster) {
-        this.cluster = cluster;
+    void setLocalMember(Member member) {
+        this.localMember = fromMember(member);
+
+        this.members.put(localMember.name(), localMember);
     }
 
     /**
      * Delegates the received topology event to the registered event handlers.
      */
-    void fireEvent(MembershipEvent event) {
+    void onMembershipEvent(MembershipEvent event) {
         ClusterNode member = fromMember(event.member());
         for (TopologyEventHandler handler : getEventHandlers()) {
             switch (event.type()) {
                 case ADDED:
+                    members.put(member.name(), member);
+
                     handler.onAppeared(member);
+
                     break;
+
                 case LEAVING:
                 case REMOVED:
```

Comment:
maybe we should start treating these events separately? Currently `onDisappeared` is fired twice, which I don't think is entirely correct

Reply:
I suppose we can examine the map and look whether it still contains disappearing member (in case if node was not stopped gracefully)